### PR TITLE
getfileinfo for dirs and appveyor testing

### PR DESCRIPTION
**Pull request by efiop**
Signed-off-by: Ruslan Kuprieiev <kupruser@gmail.com>

### CODE REVIEW

**Comment by sid0** [comment]
Thanks so much for the PR! I don't have a Windows build environment at the moment, but I'll try to get one set up to look at this some time later this week.

**Comment by efiop** [comment]
@sid0 Hey, if you could enable appveyor(https://www.appveyor.com/ , it is free for opensource and should only take you a few minutes to provide access for it) for your repo, I could set it up myself, so that you have a CI running and testing this stuff. I am using your extremely useful project in ours(dataversioncontrol/dvc) and thus very much interested in ntfsutils being tested and stable.

**Comment by sid0** [comment]
Hey @efiop -- sorry about the delay. I've been intermittent for personal health reasons.

Good point re appveyor! I use it for a couple of other projects. Would you like to set it up for this? I'm happy to give you access to manage the Appveyor installation if you give me your AppVeyor email address. Thanks!

**Comment by efiop** [comment]
@sid0 Sorry, didn't mean to hurry you. Hope all is well.

Sure, I am happy to set it up. My email on appveyor is the same as in my github profile(it should be visible at github.com/efiop).

**Comment by sid0** [comment]
@efiop just added you! Let me know how it goes

**Comment by efiop** [comment]
@sid0 Thank you! Already working on it, should be ready today.

**Comment by efiop** [comment]
@sid0 I've changed the original patch a bit, so it only adds FILE_FLAG_BACKUP_SEMANTICS if path is pointing to the directory. I also added a few simple tests to cover hardlinks and junctions and setup appveyor to run those tests for python 2 and 3. Please take a look when you have time. Thank you :)

File: ntfsutils/fs.py [hunk 1]
```diff
@@ -90,7 +91,12 @@ def getfileinfo(path):
     Return information for the file at the given path. This is going to be a
     struct of type BY_HANDLE_FILE_INFORMATION.
     """
-    hfile = CreateFile(path, GENERIC_READ, FILE_SHARE_READ, None, OPEN_EXISTING, 0, None)
+    if os.path.isdir(path):
```
Comment (sid0) [inline ntfsutils/fs.py]: So I'm a little bit concerned about [TOCTOU](https://en.wikipedia.org/wiki/Time_of_check_to_time_of_use) races here -- i.e. if something comes in and changes the path in between the `os.path.isdir` call and the `CreateFile` below. Would it be worth doing something like trying to call `CreateFile` without this flag first, then if it fails (and you can see that it's a directory) to try and call it with this flag?

Comment (efiop) [reply ntfsutils/fs.py]: I totally agree. Updated the code to try once again with FILE_FLAG_BACKUP_SEMANTICS flag. Please take a look.

**Comment by efiop** [comment]
On the other hand, maybe it is worth simply creating a getdirinfo() and leaving getfileinfo() intact. What do you think?

**Comment by sid0** [comment]
Yeah, I like the idea to create `getdirinfo()` -- if the user's unsure whether it's a file or a directory they can call both themselves.

**Comment by efiop** [comment]
Agreed :+1: . Please see getdirinfo() implemented in the new patch version.

File: ntfsutils/fs.py [hunk 2]
```diff
@@ -85,12 +85,8 @@ class BY_HANDLE_FILE_INFORMATION(ctypes.Structure):
 CloseHandle.argtypes = [HANDLE]
 CloseHandle.restype = BOOL
 
-def getfileinfo(path):
-    """
-    Return information for the file at the given path. This is going to be a
-    struct of type BY_HANDLE_FILE_INFORMATION.
-    """
-    hfile = CreateFile(path, GENERIC_READ, FILE_SHARE_READ, None, OPEN_EXISTING, 0, None)
+def _getfileinfo(path, flags=0):
```
Comment (sid0) [inline ntfsutils/fs.py]: small nit: I don't think a default argument is very useful here -- could you have each caller pass in the flags explicitly? That would be a little clearer.

Other than that this looks great. Thank you so much!

Comment (efiop) [reply ntfsutils/fs.py]: Sure, will update in a second. Thank you!

**Comment by sid0** [comment]
Woo hoo, awesome!

**Comment by efiop** [comment]
@sid0 :tada: Thank you!

**Comment by efiop** [comment]
@sid0 Ah, one more thing. Could you please release this version, so it is available through pip? No rush though.

**Comment by sid0** [comment]
Just released 0.1.4. Thanks again for your patience.

**Comment by efiop** [comment]
@sid0 wow, that was fast! Thank you!

**Pull request merged**